numpy/ma/core.py

```diff
@@ -3190,20 +3195,31 @@ def astype(self, newtype):
          [7 -- 9]]
 
         """
-        newtype = np.dtype(newtype)
-        newmasktype = make_mask_descr(newtype)
+        output = super(MaskedArray, self).astype(dtype,
+            order=order, casting=casting, subok=subok, copy=copy)
+
+        # if no copy was made, there's nothing to do
+        if output is self:
+            return self
+
+        # asked to return a base array, so nothing to decorate
+        if not subok:
+            return output
 
-        output = self._data.astype(newtype).view(type(self))
         output._update_from(self)
 
         if self._mask is nomask:
             output._mask = nomask
         else:
-            output._mask = self._mask.astype(newmasktype)
+            # we know we copied the data here, so copy the mask too
+            output._mask = self._mask.astype(
+                make_mask_descr(output.dtype),
+                order=order, casting=casting, subok=subok, copy=True
```

Comment:
By this point I think we are guaranteed to have `subok=True`, and `copy=True` is the default. So the last args are superfluous, but do no harm.

Also, (I'm not sure if this matters), but perhaps subok should be False for the mask. Eg, in `masked_matrix`s the mask is always a plain ndarray. But I guess if the rest of MA code guarantees the mask is a plain ndarray there is no harm in subok=True here.